src/app/core/internal/impl/ComponentInfo.ts

```diff
@@ -0,0 +1,117 @@
+import {Schema} from "core/schema";
+
+
+export interface PortConfig {
+    counts: Record<string, number>;
+}
+
+// Describes legal component configurations.
+// NOTE: The configuration wherein each port group has zero ports is IMPLICITLY LEGAL, because the
+//  component is UNREACHABLE in any propagator.
+export interface ComponentInfo {
+    readonly kind: string;
+    readonly portGroups: string[];
+    readonly defaultPortConfig: PortConfig;
+
+    isValidPortConfig(p: PortConfig): boolean;
+    makePortsForConfig(id: Schema.GUID, p: PortConfig): Schema.Port[] | undefined;
+
+    // i.e. Prevents fan-in on digital ports, could prevent illegal self-connectivity
+    isValidPortConnectivity(wires: Map<Schema.Port, Schema.Port[]>): boolean;
+
+    // Also i.e. thumbnail, display name, description, etc.
+}
+
+export interface ComponentInfoProvider {
+    get(kind: string): ComponentInfo | undefined;
+}
+
+// TODO: seems kind of heavy
+export function IsValidPortList(info: ComponentInfo, ports: Schema.Port[]): boolean {
+    const countMap = new Map<string, number>();
+    ports.forEach(({ group }) =>
+        countMap.set(group, (countMap.get(group) ?? 0) + 1));
+
+    const counts: Record<string, number> = {};
+    info.portGroups.forEach((g) =>
+        { counts[g] = countMap.get(g) ?? -1});
+    return info.isValidPortConfig({ counts });
```

Comment:
Mmm I think a more straightforward way would be
```ts
// Get initial counts for each defined port group
const counts = info.portGroups.reduce((g, counts) => ({ ...counts, [g]: 0 }), {});
if (ports.some((group) => !(group in counts)))
    return false;
// Count the ports
ports.forEach(({ group }) => counts[group]++);
return info.isValidPortConfig({ counts });
```

Reply:
i don't think the original or the new one does the right thing.  `counts` should contain all groups in `ports`, even if they don't exist in `info`.  `isValidPortConfig` already needs to accept arbitrary `PortConfig`s.  This does the thing I was intending:

```typescript
    // Count ports in each group
    const countMap = new Map<string, number>();
    ports.forEach(({ group }) =>
        countMap.set(group, (countMap.get(group) ?? 0) + 1));

    // Convert map to record
    const counts = {} as Record<string, number>;
    countMap.forEach((v, k) => counts[k] = v);

    return info.isValidPortConfig({ counts });
```

Reply:
Okay then I mean it's even easier:
```ts
const counts = {} as Record<string, number>;
ports.forEach(({ group }) =>
    (counts[group] = (counts[group ?? 0) + 1));
return info.isValidPortConfig({ counts });
```
The main point being that the map is unnecessary